src/outputs/restart_hdf5.hpp

```diff
@@ -0,0 +1,256 @@
+//========================================================================================
+// Parthenon performance portable AMR framework
+// Copyright(C) 2020-2022 The Parthenon collaboration
+// Licensed under the 3-clause BSD License, see LICENSE file for details
+//========================================================================================
+// (C) (or copyright) 2020-2021. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef OUTPUTS_RESTART_HDF5_HPP_
+#define OUTPUTS_RESTART_HDF5_HPP_
+//! \file io_wrapper.hpp
+//  \brief defines a set of small wrapper functions for MPI versus Serial Output.
+
+#include <string>
+#include <vector>
+
+#include "config.hpp"
+#include "outputs/restart.hpp"
+#ifdef ENABLE_HDF5
+#include <hdf5.h>
+
+#include "interface/metadata.hpp"
+#include "outputs/parthenon_hdf5_types.hpp"
+
+using namespace parthenon::HDF5;
+// TODO(someone) the following "else" is very ugly but fixes missing types when not
+// using hdf5. The issue is that some of the restart machinery lives outside of
+// restart.cpp and restart.hpp and uses the interfaces below. We should overall update
+// the logic to reduce the ifdefs for HDF5 in this (and other restart related) files.
```

Comment:
gross. Maybe we should move these definitions to the `parthenon_hdf5_types.hpp` file? Then that file can always be included where relevant?

Reply:
Let me double check.
This PR might actually have resolved the issue as what's need in other places is just the knowledge of the abstract class, which does not need hdf5.
The only item left might be writing Params, but that could potentially be alleviated by adding a `DumpParams` interface that accepts the package as paramter rather than having a `DumpToHdf5` call living inside params.

Reply:
Good news. Given that the `DatasetHandle` (and all associate calls) are now private to the hdf5 reader, the `else` statement could be removed.